File: apps/lark-server/src/middleware/metrics.ts
import { Counter, Histogram, Gauge, Registry, collectDefaultMetrics } from 'prom-client';
import type { Context, Next } from 'hono';
import { Hono } from 'hono';

export const register = new Registry();
collectDefaultMetrics({ register });

const httpRequestsTotal = new Counter({
    name: 'http_requests_total',
    help: 'Total HTTP requests',
    labelNames: ['method', 'path', 'status'] as const,
    registers: [register],
});

const httpRequestDuration = new Histogram({
    name: 'http_request_duration_seconds',
    help: 'HTTP request duration in seconds',
    labelNames: ['method', 'path'] as const,
    registers: [register],
});

const httpRequestsInFlight = new Gauge({
    name: 'http_requests_in_flight',
    help: 'Number of HTTP requests currently being processed',
    registers: [register],
});

/**
 * Hono middleware that records Prometheus HTTP metrics.
 */
export async function metricsMiddleware(c: Context, next: Next): Promise<void> {
    httpRequestsInFlight.inc();
    const start = performance.now();
    try {
        await next();
    } finally {
        const duration = (performance.now() - start) / 1000;
        httpRequestsInFlight.dec();
        httpRequestsTotal.inc({ method: c.req.method, path: c.req.path, status: String(c.res.status) });
        httpRequestDuration.observe({ method: c.req.method, path: c.req.path }, duration);
    }
}

/**
 * Sub-app with /metrics endpoint.
 */
export const metricsApp = new Hono();
metricsApp.get('/metrics', async (c) => {
    return c.text(await register.metrics(), 200, { 'Content-Type': register.contentType });
});

File: apps/agent-service/app/middleware/metrics.py
"""Prometheus metrics middleware for FastAPI (ASGI)."""

import time

from prometheus_client import Counter, Gauge, Histogram, generate_latest
from starlette.requests import Request
from starlette.responses import Response
from starlette.types import ASGIApp, Receive, Scope, Send

REQUEST_COUNT = Counter(
    "http_requests_total",
    "Total HTTP requests",
    ["method", "path", "status"],
)
REQUEST_DURATION = Histogram(
    "http_request_duration_seconds",
    "HTTP request duration in seconds",
    ["method", "path"],
)
REQUEST_IN_FLIGHT = Gauge(
    "http_requests_in_flight",
    "Number of HTTP requests currently being processed",
)


class PrometheusMiddleware:
    """ASGI middleware that records HTTP metrics."""

    def __init__(self, app: ASGIApp) -> None:
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        request = Request(scope)

        # Serve /metrics endpoint directly
        if request.url.path == "/metrics":
            body = generate_latest()
            response = Response(
                body, media_type="text/plain; version=0.0.4; charset=utf-8"
            )
            await response(scope, receive, send)
            return

        path = request.url.path
        method = request.method
        status_code = 500

        # Capture status code from response
        async def send_wrapper(message):
            nonlocal status_code
            if message["type"] == "http.response.start":
                status_code = message["status"]
            await send(message)

        REQUEST_IN_FLIGHT.inc()
        start = time.monotonic()
        try:
            await self.app(scope, receive, send_wrapper)
        finally:
            duration = time.monotonic() - start
            REQUEST_IN_FLIGHT.dec()
            REQUEST_COUNT.labels(
                method=method, path=path, status=str(status_code)
            ).inc()
            REQUEST_DURATION.labels(method=method, path=path).observe(duration)
